src/store/reducers/Cache.ts

```diff
@@ -0,0 +1,70 @@
+import { WithId } from "utils/id";
+import { UserWithLocation } from "types/User";
+import firebase from "firebase/app";
+
+import { CacheActions, CacheActionTypes } from "store/actions/Cache";
+
+export type CacheState = {
+  usersRecord: Record<string, WithId<UserWithLocation>>;
+  usersArray: WithId<UserWithLocation>[];
+};
+
+const initialState: CacheState = {
+  usersRecord: {},
+  usersArray: [],
+};
+
+export const cacheReducer = (
+  state = initialState,
+  action: CacheActions
+): CacheState => {
+  switch (action.type) {
+    case CacheActionTypes.RELOAD_USER_CACHE:
+      console.log("reloading user cache");
+      firebase
+        .firestore()
+        .collection("users")
+        .get()
+        .then((snapshot) => {
+          const usersRecord = {} as Record<string, WithId<UserWithLocation>>;
+
+          snapshot.forEach((doc: firebase.firestore.QueryDocumentSnapshot) => {
+            const user: WithId<UserWithLocation> = doc.data() as WithId<UserWithLocation>;
+            user.id = doc.id;
+            usersRecord[doc.id] = user;
+          });
+
+          const usersArray = Object.values(usersRecord);
+
+          state.usersArray = usersArray;
+          state.usersRecord = usersRecord;
+
+          console.log(snapshot.docs);
+          // const data = snapshot.map((doc: firebase.firestore.QueryDocumentSnapshot) => doc.data())
+          // state.usersArray = snapshot.docs as unknown as WithId<UserWithLocation>[]
+          // state.usersRecord = snapshot.docs.reduce((acc, user) => {
+          //   acc[user.id as string]: user as unknown as WithId<UserWithLocation>
+          // }) as unknown as Record<string, WithId<UserWithLocation>>
+        });
+
+      //not the state that will be update... do async operation above outside and call in with the list of user instead
+      return state;
+
+    case CacheActionTypes.UPDATE_USER_CACHE:
+      return {
+        ...state,
+        usersRecord: {
+          ...state.usersRecord,
+          [action.payload.id]: action.payload.user,
+        },
+        usersArray: state.usersArray.map((u) => {
+          //console.log(u.id);
+
+          if (u.id !== action.payload.id) return u; //don't change
+          return action.payload.user;
+        }),
+      };
+    default:
+      return state;
+  }
+};
```

Comment:
Here's the reducer logic

Reply:
This query will load literally all users within the user collection every single time we poll, regardless of whether those users have changed or not:

```typescript
firebase
  .firestore()
  .collection("users")
  .get()
  .then((snapshot) => {
```

Firebase queries allow us to only fetch data that has changed since X time, and/or 'after' a certain snapshot, etc. This would dramatically reduce the amount of data that needs to be loaded each time, and thus will reduce bandwidth usage + also reduce the cost of reads from firebase.

Reply:
It might be possible to use query cursors or similar https://firebase.google.com/docs/firestore/query-data/query-cursors ?

Reply:
@denisdimitrov is going to look into this and assess our options here

Reply:
> It might be possible to use query cursors or similar https://firebase.google.com/docs/firestore/query-data/query-cursors ?

@0xdevalias @mike-lvov Looks like this is mainly used for pagination purposes and won't do the job. It's not flexible enough and doesn't provide the necessary options. Basically, it allows you to add "offset" and "limit" to your query.

I think we should be providing the world users to this action and the reducer should be using only those users and cache their locations. The list of user will be less that fetching all users but it still doesn't change the fact nor fix the problem that unnecessary users will be cache without actual change in their location. I will keep investigating and reading about other possible solutions.

Looks like there are value events that might be useful, currently reading about them: https://firebase.google.com/docs/database/admin/retrieve-data#section-event-types

Reply:
Mm, I wasn't sure if that link I provided specifically would work, but yeah, makes sense as to why it wouldn't as you said.

My original thought (not backed up by any specific links) was to just query for 'give me things that have changed since I last queried'. So I assumed there may be a timestamp type of metadata thing we could potentially query on or similar.

Reply:
> Looks like there are value events that might be useful, currently reading about them: https://firebase.google.com/docs/database/admin/retrieve-data#section-event-types

@denisdimitrov Which particular part of this? That's for 'realtime database' not firestore, but it also just seems to be equivalent to our standard 'snapshot listener' connection that this PR rips out due to the issues we were having with it.

Reply:
> @denisdimitrov Which particular part of this? That's for 'realtime database' not firestore, but it also just seems to be equivalent to our standard 'snapshot listener' connection that this PR rips out due to the issues we were having with it.

@0xdevalias You are right, this won't  do the job either. 

I have an idea but I might be talking BS. We can fetch world users once (by their user locations) and then use a to populate an array with the changed users, that way the next time the interval triggers it will have to fetch and update only that list of users. Wdyt?

Reply:
> and then use a to populate an array with the changed users,

Is this missing something? How will you tell it to fetch only a list of users? And how will you know which ones to query?

On this environment basically every user that exists is going to be a 'related user' anyway, so this wouldn't reduce the number we need to get I don't believe.

Reply:
Yup, there was a typo. We can use a listener, onSnapshot or something like that which can populate the user id in the new array, and then we fetch only the users from that array, by bulk id query.

Reply:
With bulk id query we can only do 10 per query I believe, so that would be 100+ queries based on the number of users on the environment.

What is the snapshot/userIds being used for here? Just to detect users which have had updates? The snapshot sends the whole document each time anyway, so this would be creating extra double work, and I don't believe would solve the issue we're having at all.

Reply:
In my 'proper fix' PR i'm looking into writing a 'queued throttle' queue/cache to batch snapshot updates together, and then process them every X timeframe; since my observation is that it's the 'constant drip' of updates that causes more of the issue than making a bunch of changes at once.

Reply:
>With bulk id query we can only do 10 per query I believe, so that would be 100+ queries based on the number of users on the environment.

That's true, but it will run only once every minute. It should be much better than live updates.

>What is the snapshot/userIds being used for here? Just to detect users which have had updates?

Yes, that's the idea

> The snapshot sends the whole document each time anyway, so this would be creating extra double work, and I don't believe would solve the issue we're having at all.

Fair enough. I'll search for something else.

P.S. Looking at your WIP fix PR atm: https://github.com/sparkletown/sparkle/pull/1833/files?file-filters%5B%5D=.ts#diff-ae206d6bb7c21f941d94450ed2c3e15ec38e85d6898407bfe91550fadc78b6cbR117-R172
Will let you know if I think of something else.

Reply:
You can see the overview I wrote about my 'queued batcher' commit here: https://github.com/sparkletown/sparkle/pull/1833#pullrequestreview-707134935